prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -22,54 +24,108 @@ def floatify(value, default=float("nan")):
+        grouper_key = (generator["interconnect"], generator["state"])
+        _, array_index = state_trees[grouper_key].query(generator["xyz"])
+        sub_index = subs_state_lookup[grouper_key][array_index]
+        return sub_index
+
+    # Translate lat/lon to 3D positions (assume spherical earth, origin at center)
+    substations_with_xyz = substations.assign(
+        xyz=substations.apply(lambda x: latlon_to_xyz(x["lat"], x["lon"]), axis=1)
+    )
+    generators_with_xyz = generators.assign(
+        xyz=generators.apply(
+            lambda x: (
+                pd.NA
+                if pd.isna(x["lat"]) or pd.isna(x["lon"])
+                else latlon_to_xyz(x["lat"], x["lon"])
+            ),
+            axis=1,
+        )
+    )
+
+    # Group substations by state to build KDTrees
+    subs_state_lookup = substations_with_xyz.groupby(["interconnect", "STATE"]).groups
+    # Group substations by ZIP code for a fallback for generators without coordinates
+    subs_zip_groupby = substations_with_xyz.groupby(["interconnect", "ZIP"])
+
+    # Create a KDTree for each combination of state and interconnect
+    state_trees = {
+        key: KDTree(np.array(substations_with_xyz.loc[sub_ids, "xyz"].tolist()))
+        for key, sub_ids in subs_state_lookup.items()
+    }
+    # Ensure that we have a tree for every generator
+    gens_state_groupby = generators_with_xyz.groupby(["interconnect", "state"])
+    missing_groups = set(gens_state_groupby.groups) - set(state_trees)
+    if len(missing_groups) > 0:
+        # There are some combinations of generator (interconnect, state) without subs
+        allowable_border_states = {"KS", "NE", "OK"}
+        for interconnect, state in missing_groups:
+            if state in allowable_border_states:
+                # Assume that the interconnect and state are correct
+                print(
+                    f"no substations within ({interconnect}, {state}), "
+                    f"will map generators to substations within {interconnect} instead"
+                )
+                # Find all substations for the interconnection
+                new_subs = substations_with_xyz.query("interconnect == @interconnect")
+                # Extend the 'true' combinations of (interconnect, state) with fakes
+                state_trees[(interconnect, state)] = KDTree(
+                    np.array(new_subs["xyz"].tolist())
+                )
+                subs_state_lookup[(interconnect, state)] = new_subs.index
+            else:
+                # Assume that the state is correct, the interconnect is wrong
+                print(
+                    f"no substations within ({interconnect}, {state}), "
+                    f"will map generators to substations within {state} instead"
+                )
+                (assumed,) = {
+                    interconnect
+                    for interconnect, state_list in const.interconnect2state.items()
```

Comment:
We have `ignore` and `split` as keys in `const.interconnect2state`. Reading the std outputs, it doesn't seem to matter given generators in those states can find corresponding substations.